doc/user/flowspec.rst

```diff
@@ -0,0 +1,330 @@
+- Misc layer 4 TCP flags.
+
+A combination of the above rules is applied for traffic filtering. This is encoded
+as part of specific BGP extended communities and the action can range from the
+obvious rerouting (to nexthop or to separate VRF) to shaping, or discard.
+
+Following IETF drafts and RFCs have been used to implement FRR Flowspec:
+
+- [RFC5575]_
+- [Draft IETF IDR Flowspec redirect IP]_
+
+
+.. _design-principles-flowspec:
+
+
+Design Principles
+-----------------
+
+FRR implements the Flowspec client side, that is to say that BGP is able to
+receive Flowspec entries, but is not able to act as manager and send Flowspec entries.
+
+FRRouting uses policy based routing mechanisms to apply Flowspec configuration to the
+underlying system. The underlying system is Linux, and provides a rich enough API to
+implement this policy routing.
+
+- by filtering the traffic based on `Linux NetFilter`.
+
+`Linux NetFilter` provides a set of tools like ipset and iptables that are powerful
+enough to be able to filter such flowspec filter rule.
+
+- by using non standard routing tables (through ip rule tool from `Linux Iproute2`).
+
+That tool is already used by pbrd daemon that provides basic routing based on IP source,
+and IP destination criterium.
+
+Below example is an illustration of what BGP flowspec will inject in the underlying
+system:
+
+.. code-block:: shell
+
+   # linux shell
+   ipset create match0x102 hash:net,net counters
+   ipset add match0x102 32.0.0.0/16,40.0.0.0/16
+   iptables -N match0x102 -t mangle
+   iptables -A match0x102 -t mangle -j MARK --set-mark 102
+   iptables -A match0x102 -t mangle -j ACCEPT
+   iptables -i ntfp3 -t mangle -I PREROUTING -m set --match-set match0x102 src,dst
+                -g match0x102
+   ip rule add fwmark 102 lookup 102
+   ip route add 40.0.0.0/16 via 44.0.0.2 table 102
+
+For handling an incoming Flowspec entry, the following workflow is applied:
+
+- incoming Flowspec entries are handled by `bgpd`, and are stored in the BGP RIB.
+- Flowspec entry will be installed according to its complexity.
+
+It will be installed if one of the following filtering action is seen on the BGP
+extended community: either redirect IP, or redirect VRF, in conjunction with rate
+option, for redirecting traffic. Or rate option set to 0, for discarding traffic.
+
+According to the degree of complexity of the Flowspec entry, it will be installed
+in *zebra* RIB. For more information about what is supported in the FRR implementation
+as rule, see `Limitations / Known Issues` chapter. Flowspec entry is split inseveral
+parts before being sent to *zebra*.
+
+- *zebra* daemon is receiving the policy routing configuration
+
+Policy-routing entities necessary to policy route the traffic in the underlying system,
+is received by *zebra*. Two filtering contexts will be created or appended (`Linux
+NetFilter ipset` and `Linux netfilter iptable` entry). If the traffic has to be dropped,
```

Comment:
I'm not sure what the appropriate markup is here since this reads like a combination of command syntax and the name of a program, but I'm pretty sure `Linux NetFilter ipset` could stand to be clarified. Also please consistently capitalize Netfilter as `Netfilter`.

Reply:
ok